ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PointSchedule/Shrinking.hs

```diff
@@ -0,0 +1,96 @@
+                     PeerSchedule, peerSchedulesBlocks)
+import           Test.Consensus.PointSchedule.Peers (Peers (..))
+import           Test.QuickCheck (shrinkList)
+import           Test.Util.TestBlock (TestBlock, TestHash (unTestHash))
+
+-- | Shrink a 'Peers PeerSchedule'. This does not affect the honest peer; it
+-- does, however, attempt to remove other peers or ticks of other peers. The
+-- block tree is trimmed to keep only parts that are necessary for the shrunk
+-- schedule.
+shrinkPeerSchedules ::
+  GenesisTest ->
+  Peers PeerSchedule ->
+  StateView ->
+  [(GenesisTest, Peers PeerSchedule)]
+shrinkPeerSchedules genesisTest schedule _stateView =
+  shrinkOtherPeers shrinkPeerSchedule schedule <&> \shrunkSchedule ->
+    let trimmedBlockTree = trimBlockTree' shrunkSchedule (gtBlockTree genesisTest)
+     in (genesisTest{gtBlockTree = trimmedBlockTree}, shrunkSchedule)
+
+-- | Shrink a 'PeerSchedule' by removing ticks from it. The other ticks are kept
+-- unchanged.
+shrinkPeerSchedule :: PeerSchedule -> [PeerSchedule]
+shrinkPeerSchedule = shrinkList (const [])
+
+-- | Shrink the 'others' field of a 'Peers' structure by attempting to remove
+-- peers or by shrinking their values using the given shrinking function.
+shrinkOtherPeers :: (a -> [a]) -> Peers a -> [Peers a]
+shrinkOtherPeers shrink Peers{honest, others} =
+  map (Peers honest . Map.fromList) $
+    shrinkList (traverse (traverse shrink)) $ Map.toList others
+
+-- | Remove blocks from the given block tree that are not necessary for the
+-- given peer schedules. If entire branches are unused, they are removed. If the
+-- trunk is unused, then it remains as an empty anchored fragment.
+trimBlockTree' :: Peers PeerSchedule -> BlockTree TestBlock -> BlockTree TestBlock
+trimBlockTree' = keepOnlyAncestorsOf . peerSchedulesBlocks
+
+-- | Given some blocks and a block tree, keep only the prefix of the block tree
+-- that contains ancestors of the given blocks.
+keepOnlyAncestorsOf :: [TestBlock] -> BlockTree TestBlock -> BlockTree TestBlock
+keepOnlyAncestorsOf blocks bt =
+    let leafs = blocksWithoutDescendents blocks
+        trunk = keepOnlyAncestorsOf' leafs (btTrunk bt)
+        branches = mapMaybe (fragmentToMaybe . keepOnlyAncestorsOf' leafs . btbSuffix) (btBranches bt)
+     in foldr addBranch' (mkTrunk trunk) branches
+  where
+    fragmentToMaybe (Empty _) = Nothing
+    fragmentToMaybe fragment  = Just fragment
+
+    -- | Given some blocks and a fragment, keep only the prefix of the fragment
+    -- that contains ancestors of the given blocks.
+    keepOnlyAncestorsOf' :: [TestBlock] -> AnchoredFragment TestBlock -> AnchoredFragment TestBlock
+    keepOnlyAncestorsOf' leafs = takeWhileOldest (\block -> (block `isAncestorOf`) `any` leafs)
+
+    -- | Return a subset of the given blocks containing only the ones that do
+    -- not have any other descendents in the set.
+    blocksWithoutDescendents :: [TestBlock] -> [TestBlock]
+    blocksWithoutDescendents =
```

Comment:
`blocksWithoutDescendents` is similar to the unoptimized

```haskell
blocksWithoutDescendents blocks =
  filter (\b -> all (not . (`isDescendentOf` b)) blocks) blocks
```

Given that both the current and the unoptimized implementations are quadratic, it would be good to consider if it is worth making the maintainer deal with the larger implementation.

I thought I had commented this last Friday, but I somehow failed at it.

Reply:
The only semantic difference I can see is in the handling of duplicates where, I think, your version will just drop them. But we could very much remove duplicates while sorting or something. I think that's a nice idea, it would make the code much more readable. The gain in performance is not much and also we don't care about performances that much in shrinking as long as it's not crazy long.

Reply:
Wait, it's not only a matter of duplicates, a block is ancestor of itself so this comprehension will always be empty. We could obtain:
```haskell
blocksWithoutDescendents bs =
  [ b | b <- nubOrd bs, not ((b `isStrictAncestorOf`) `any` bs) ]
```
but it starts being a bit more convoluted.

Reply:
Also, the current implementation is more `O(nm⋅ln(n))` with `n` the size of the list and `m` the number of branches in the corresponding block tree, which in theory is `O(n²⋅ln(n))` but in practice for our use case `m` tends to be quite tiny compared to `n`. So I think I prefer to keep the current version.

Reply:
Ah, so `blocksWithoutDescendents` is also removing duplicates. That would be worth pointing out in a comment. Slight improvement upon your last definition:

```haskell
blocksWithoutDescendents blocks =
  let bs = nubOrd blocks
   in [ b | b <- bs, not ((b `isStrictAncestorOf`) `any` bs) ]
```

Reply:
It removes the duplicates but that is more of a side effect and I don't think it really matters for the correction of how we use it. Actually, my comments about the duplicates were more about me getting confused. The real problem was the fact that `isAncestorOf` also returns true on equal blocks and therefore your first suggestion would always return the empty list because the block `b` would always be in `bs`. I think it is after all fine to only use:
```haskell
blocksWithoutDescendents bs =
  [ b | b <- bs, not ((b `isStrictAncestorOf`) `any` bs) ]
```
without removal of duplicates but with strict ancestry check. Do you prefer this version over `foldl` despite my comment above on complexity?

Reply:
I hadn't noticed the amount of branches would affect the complexity. I think it is fine to keep the current implementation.

Probably it will be helpful to future programmers if some of the points in this conversation are written in a comment.

Reply:
After a night of sleep, I now prefer the list comprehension. Done in d37e976d45ccb0e80fa15e11a7d2211214cbb8ab and fedb55c4cb304da697b032c7daf9e3da0f6e1251.